### PR TITLE
[CELEBORN-1574] Speed up unregister shuffle by batch processing

### PR DESCRIPTION
**Pull request by zaynt4606**
### What changes were proposed in this pull request?
In order to speed up the resource releasing，this PR Unregister shuffle in batch;

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT & Local cluster testing

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -4418,7 +4419,7 @@ object CelebornConf extends Logging {
       .version("0.3.0")
       .doc("Interval for client to check expired shuffles.")
       .timeConf(TimeUnit.MILLISECONDS)
-      .createWithDefaultString("60s")
+      .createWithDefaultString("20s")
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4422]: Why is the default value changed? Any production practice?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -4450,6 +4451,14 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
+  val CLIENT_BATCH_REMOVE_EXPIRED_SHUFFLE: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.shuffle.batchHandleRemoveExpiredShuffles.enabled")
+      .categories("client")
+      .version("0.5.0")
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4457]:
```suggestion
      .version("0.6.0")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 1 above)
Comment (zaynt4606) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4422]: Draft need to discuss

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 3]
```diff
@@ -1655,6 +1693,25 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
+    try {
+      logInfo(s"AskSync UnregisterShuffle for ${message.getShuffleIdsList}")
+      masterClient.askSync[PbBatchUnregisterShuffleResponses](
+        message,
+        classOf[PbBatchUnregisterShuffleResponses])
+    } catch {
+      case e: Exception =>
+        logError(s"AskSync UnregisterShuffle for ${message.getShuffleIdsList} failed.", e)
+        val map = JavaUtils.newConcurrentHashMap[Integer, StatusCode]()
+        val shuffleIds = message.getShuffleIdsList
```
Comment (mridulm) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1707]: Only for subset where `PbUnregisterShuffleResponsesInfo.status` is failed ?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 1 above)
Comment (mridulm) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4422]: Agree, let us keep the defaults unless there are good reasons to change it.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -4450,6 +4451,14 @@ object CelebornConf extends Logging {
...
+      .doc("This is an optimization on remove Expired Shuffles and it's true by default.")
+      .booleanConf
+      .createWithDefault(true)
```
Comment (mridulm) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4460]: Let us start with `false` and flip it in later releases.

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java [hunk 5]
```diff
@@ -118,6 +118,16 @@ public ResourceResponse handleWriteRequest(ResourceProtos.ResourceRequest reques
           metaSystem.updateUnregisterShuffleMeta(shuffleKey);
           break;
 
+        case BatchUnRegisterShuffle:
+          List<String> shuffleKeys =
+              request.getBatchUnregisterShuffleRequest().getShuffleKeysList();
+          shuffleKeys.forEach(
+              key -> {
+                LOG.debug("Handle unregister shuffle for {}", key);
```
Comment (mridulm) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java:126]: pull this out of the foreach and dump all in one shot ? debug logging is already noisy :-)

**Review by mridulm — COMMENTED** [review]
Took a quick look - this is promising, thanks for contributing it !

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HAMasterMetaManager.java [hunk 6]
```diff
@@ -109,6 +109,24 @@ public void handleUnRegisterShuffle(String shuffleKey, String requestId) {
...
+          ResourceRequest.newBuilder()
+              .setCmdType(Type.BatchUnRegisterShuffle)
+              .setRequestId(requestId)
+              .setBatchUnregisterShuffleRequest(
+                  ResourceProtos.BatchUnregisterShuffleRequest.newBuilder()
+                      .addAllShuffleKeys(shuffleKeys)
+                      .build())
+              .build());
+    } catch (CelebornRuntimeException e) {
+      LOG.error("Handle unregister shuffle for {} failed!", shuffleKeys, e);
```
Comment (SteNicholas) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HAMasterMetaManager.java:125]:
```suggestion
      LOG.error("Batch handle unregister shuffle for {} failed!", shuffleKeys, e);
```

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java (hunk 5 above)
Comment (SteNicholas) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java:126]: +1.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 7]
```diff
@@ -1583,6 +1590,37 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     }
   }
 
+  private def removeBatchExpiredShuffles(): Unit = {
+    val shuffleIds = new util.ArrayList[Integer]()
+    unregisterShuffleTime.keys().asScala.foreach { shuffleId =>
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1595]: Thank you for the code review!
I think it would be beneficial to focus solely on the batch processing in this pull request. This way, we can maintain the same time interval and condition as before.
Thanks!

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java (hunk 5 above)
Comment (zaynt4606) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java:126]: ok thanks！

File: tests/spark-it/src/test/scala/org/apache/celeborn/tests/client/LifecycleManagerUnregisterShuffleSuite.scala [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+  test("test unregister shuffle in batch") {
+    val conf = celebornConf.clone
+    val lifecycleManager: LifecycleManager = new LifecycleManager(APP, conf)
+    val ids = new util.ArrayList[Integer](10)
+    val shuffleIds = new util.ArrayList[Integer](10)
+    0 until 10 foreach { num =>
+      ids.add(num)
+      shuffleIds.add(num + 1)
+    }
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply tests/spark-it/src/test/scala/org/apache/celeborn/tests/client/LifecycleManagerUnregisterShuffleSuite.scala:55]: Thanks！

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala (hunk 3 above)
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1693]: In this case, all subsets are anomalous and there is no need for repeated checks.The part related to filling the map has been removed.

File: tests/spark-it/src/test/scala/org/apache/celeborn/tests/client/LifecycleManagerUnregisterShuffleSuite.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (mridulm) [reply tests/spark-it/src/test/scala/org/apache/celeborn/tests/client/LifecycleManagerUnregisterShuffleSuite.scala:58]: With this change proposed, we can also avoid all the `asScala` calls below :-)

**Review by mridulm — APPROVED** [review]
Looks good to me, there is a pending comment by @bowenliang123 [here](https://github.com/apache/celeborn/pull/2701#discussion_r1730852948) which is useful as well.

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 9]
```diff
@@ -978,6 +988,25 @@ private[celeborn] class Master(
...
+    shuffleIds.foreach { shuffleId =>
+      val shuffleKey = Utils.makeShuffleKey(applicationId, shuffleId)
+      shuffleKeys.add(shuffleKey)
+    }
+    statusSystem.batchHandleUnRegisterShuffles(shuffleKeys, requestId)
+    logInfo(s"Unregister shuffle $shuffleKeys")
+    shuffleIds.foreach { shuffleId =>
+      map.put(shuffleId, StatusCode.SUCCESS)
+    }
+    context.reply(BatchUnregisterShuffleResponses(StatusCode.SUCCESS, map))
```
Comment (bowenliang123) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:1007]:
```suggestion
    val shuffleKeys = shuffleIds.map(shuffleId => Utils.makeShuffleKey(applicationId, shuffleId)).asJava
    statusSystem.batchHandleUnRegisterShuffles(shuffleKeys, requestId)
    
    logInfo(s"Unregister shuffle $shuffleKeys")
    
    val map = shuffleIds.map(shuffleId => shuffleId -> StatusCode.SUCCESS).toMap.asJava
    context.reply(BatchUnregisterShuffleResponses(StatusCode.SUCCESS, map))
```
Another minor suggestion.
In a more concise Scala way for the implementation.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 10]
```diff
@@ -1573,12 +1577,32 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
+          // if unregister shuffle not success, wait next turn
+          if (StatusCode.SUCCESS == Utils.toStatusCode(unregisterShuffleResponse.getStatus)) {
+            unregisterShuffleTime.remove(shuffleId)
+          }
+        } else {
+          batchRemoveShuffleIds.add(shuffleId)
+        }
+      }
+    }
+    if (!batchRemoveShuffleIds.isEmpty) {
```
Comment (bowenliang123) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1592]:
```suggestion
    if (batchRemoveExpiredShuffles) {
```
Use the `batchRemoveExpiredShuffles` for condition would be more easy to handle and understand. `batchRemoveShuffleIds` are only collected when `batchRemoveExpiredShuffles` is set to be `true` anyway.

Comment (mridulm) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1592]: It is not necessary that there are shuffle id's to remove.
So the `!batchRemoveShuffleIds.isEmpty` is required here to prevent sending `BatchUnregisterShuffles` with no shuffle ids.

Comment (bowenliang123) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1592]: OK. But to be honest, the current implementation is far from the best.

Comment (mridulm) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1592]: I am not sure what you mean - if there are suggestions to improve it in a PR, that would be great.

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala (hunk 9 above)
Comment (zaynt4606) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:1007]: Thanks!

**Comment by zaynt4606** [comment]
Please take a look at this PR. Thanks!
cc @RexXiong @FMX

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 11]
```diff
@@ -229,6 +231,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
       new Runnable {
         override def run(): Unit = Utils.tryLogNonFatalError {
           self.send(RemoveExpiredShuffle)
+          // todo
```
Comment (FMX) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:234]: unnecessary change

File: common/src/main/proto/TransportMessages.proto [hunk 12]
```diff
@@ -38,6 +38,8 @@ enum MessageType {
   GET_REDUCER_FILE_GROUP_RESPONSE = 15;
   UNREGISTER_SHUFFLE = 16;
   UNREGISTER_SHUFFLE_RESPONSE = 17;
+  BATCH_UNREGISTER_SHUFFLES = 21;
```
Comment (FMX) [inline common/src/main/proto/TransportMessages.proto:41]: Use unused numbers to ensure these RPCs won't be treated as old RPCs.

File: common/src/main/proto/TransportMessages.proto [hunk 13]
```diff
@@ -402,10 +404,26 @@ message PbUnregisterShuffle {
...
 message PbUnregisterShuffleResponse {
   int32 status = 1;
 }
 
+message PbBatchUnregisterShuffleResponses {
+  int32 status = 1;
+  repeated PbUnregisterShuffleResponsesInfo unregisterShuffleResponsesInfo = 2;
+}
+
+message PbUnregisterShuffleResponsesInfo {
```
Comment (FMX) [inline common/src/main/proto/TransportMessages.proto:422]: I think this proto is unnecessary. Maybe 
```
message PbBatchUnregisterShuffleResponses {
  int32 status = 1;
  repeated int32 shuffleIds = 2;
}
```

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/SingleMasterMetaManager.java [hunk 14]
```diff
@@ -65,6 +65,11 @@ public void handleUnRegisterShuffle(String shuffleKey, String requestId) {
     updateUnregisterShuffleMeta(shuffleKey);
   }
 
+  @Override
+  public void batchHandleUnRegisterShuffles(List<String> shuffleKeys, String requestId) {
+    shuffleKeys.forEach(shuffleKey -> updateUnregisterShuffleMeta(shuffleKey));
```
Comment (FMX) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/SingleMasterMetaManager.java:70]:
```suggestion
    registeredShuffle.removeAll(shuffleKeys);
```

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java [hunk 15]
```diff
@@ -118,6 +118,13 @@ public ResourceResponse handleWriteRequest(ResourceProtos.ResourceRequest reques
           metaSystem.updateUnregisterShuffleMeta(shuffleKey);
           break;
 
+        case BatchUnRegisterShuffle:
+          List<String> shuffleKeys =
+              request.getBatchUnregisterShuffleRequest().getShuffleKeysList();
+          shuffleKeys.forEach(key -> metaSystem.updateUnregisterShuffleMeta(key));
```
Comment (FMX) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java:124]:
```suggestion
    metaSystem.registeredShuffle.removeAll(shuffleKeys);
```

File: master/src/main/proto/Resource.proto [hunk 16]
```diff
@@ -22,6 +22,7 @@ package deploy.master.meta;
 
 enum Type {
   UnRegisterShuffle = 12;
+  BatchUnRegisterShuffle = 28;
```
Comment (FMX) [inline master/src/main/proto/Resource.proto:25]: Place this new type to the end of this section.

File: common/src/main/proto/TransportMessages.proto (hunk 13 above)
Comment (FMX) [reply common/src/main/proto/TransportMessages.proto:422]: According to your implementation, the shuffleStatus can be ignored because there is no logic to process the failure of errors that might occur when removing the shuffle ID.

Comment (FMX) [reply common/src/main/proto/TransportMessages.proto:422]: If a batch unregistered shuffle RPC fails, retrying it can be enough.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala (hunk 11 above)
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala]: deleted! thanks!

File: common/src/main/proto/TransportMessages.proto (hunk 12 above)
Comment (zaynt4606) [reply common/src/main/proto/TransportMessages.proto]: Done!

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/SingleMasterMetaManager.java (hunk 14 above)
Comment (zaynt4606) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/SingleMasterMetaManager.java]: Done!

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java (hunk 15 above)
Comment (zaynt4606) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java]: Done!

File: master/src/main/proto/Resource.proto (hunk 16 above)
Comment (zaynt4606) [reply master/src/main/proto/Resource.proto]: Done!

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 17]
```diff
@@ -1589,10 +1592,23 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
+          }
+        } else {
+          batchRemoveShuffleIds.add(shuffleId)
+        }
+      }
+    }
+    if (!batchRemoveShuffleIds.isEmpty) {
+      val unregisterShuffleResponse = batchRequestMasterUnregisterShuffles(
+        BatchUnregisterShuffles(appUniqueId, batchRemoveShuffleIds, MasterClient.genRequestId()))
+      if (StatusCode.SUCCESS == Utils.toStatusCode(unregisterShuffleResponse.getStatus)) {
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1610]: Done~

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 18]
```diff
@@ -4476,6 +4477,14 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
+  val CLIENT_BATCH_REMOVE_EXPIRED_SHUFFLE: ConfigEntry[Boolean] =
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Done~

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 19]
```diff
@@ -4476,6 +4477,14 @@ object CelebornConf extends Logging {
...
+    buildConf("celeborn.client.shuffle.batchHandleRemoveExpiredShuffles.enabled")
+      .categories("client")
+      .version("0.6.0")
+      .doc("This is an optimization on remove Expired Shuffles and it's true by default.")
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Yes! Changed~

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 20]
```diff
@@ -1042,6 +1042,7 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
   def batchHandleChangePartitionNumThreads: Int = get(CLIENT_BATCH_HANDLE_CHANGE_PARTITION_THREADS)
   def batchHandleChangePartitionRequestInterval: Long =
     get(CLIENT_BATCH_HANDLE_CHANGE_PARTITION_INTERVAL)
+  def batchHandleRemoveExpiredShuffles: Boolean = get(CLIENT_BATCH_REMOVE_EXPIRED_SHUFFLE)
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: Done~

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java [hunk 21]
```diff
@@ -118,6 +118,13 @@ public ResourceResponse handleWriteRequest(ResourceProtos.ResourceRequest reques
           metaSystem.updateUnregisterShuffleMeta(shuffleKey);
           break;
 
+        case BatchUnRegisterShuffle:
+          List<String> shuffleKeys =
+              request.getBatchUnregisterShuffleRequest().getShuffleKeysList();
+          metaSystem.registeredShuffle.removeAll(shuffleKeys);
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java]: Done! It should use new method in AbstractMetaManager as updateUnregisterShuffleMeta ~

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/IMetadataHandler.java [hunk 22]
```diff
@@ -36,6 +36,8 @@ void handleRequestSlots(
 
   void handleUnRegisterShuffle(String shuffleKey, String requestId);
 
+  void batchHandleUnRegisterShuffles(List<String> shuffleKeys, String requestId);
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/IMetadataHandler.java]: Done!

File: master/src/test/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/RatisMasterStatusSystemSuiteJ.java [hunk 23]
```diff
@@ -844,6 +844,93 @@ public void testHandleUnRegisterShuffle() throws InterruptedException {
...
+      statusSystem.handleRequestSlots(shuffleKey, HOSTNAME1, workersToAllocate, getNewReqeustId());
+    }
+
+    Thread.sleep(3000L);
+
+    Assert.assertEquals(3, STATUSSYSTEM1.registeredShuffle.size());
+    Assert.assertEquals(3, STATUSSYSTEM2.registeredShuffle.size());
+    Assert.assertEquals(3, STATUSSYSTEM3.registeredShuffle.size());
+
+    statusSystem.batchHandleUnRegisterShuffles(shuffleKeys, getNewReqeustId());
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply master/src/test/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/RatisMasterStatusSystemSuiteJ.java]: Added these two tests in the ha mode and single mode TestSuite~

**Review by RexXiong — APPROVED** [review]
Overall LGTM, only a minor comment

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 24]
```diff
@@ -103,6 +103,8 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
   private val rpcCacheExpireTime = conf.clientRpcCacheExpireTime
   private val rpcMaxRetires = conf.clientRpcMaxRetries
 
+  private val batchRemoveExpiredShuffles = conf.batchHandleRemoveExpiredShufflesEnabled
```
[the remark this answers is not in the document]
Comment (zaynt4606) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala]: Done!

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 25]
```diff
@@ -1579,6 +1581,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
 
   private def removeExpiredShuffle(): Unit = {
     val currentTime = System.currentTimeMillis()
+    val batchRemoveShuffleIds = new util.ArrayList[Integer]()
```
Comment (SteNicholas) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1584]: Why not use `Seq`?

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 26]
```diff
@@ -1671,6 +1687,20 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     }
   }
 
+  private def batchRequestMasterUnregisterShuffles(message: PbBatchUnregisterShuffles)
+      : PbBatchUnregisterShuffleResponses = {
+    try {
+      logInfo(s"AskSync UnregisterShuffle for ${message.getShuffleIdsList}")
```
Comment (SteNicholas) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1693]:
```suggestion
      logInfo(s"AskSync BatchUnregisterShuffle for ${message.getShuffleIdsList}")
```

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 27]
```diff
@@ -1671,6 +1687,20 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
+      masterClient.askSync[PbBatchUnregisterShuffleResponses](
+        message,
+        classOf[PbBatchUnregisterShuffleResponses])
+    } catch {
+      case e: Exception =>
+        logError(s"AskSync UnregisterShuffle for ${message.getShuffleIdsList} failed.", e)
```
Comment (SteNicholas) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:1699]:
```suggestion
        logError(s"AskSync BatchUnregisterShuffle for ${message.getShuffleIdsList} failed.", e)
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 28]
```diff
@@ -4476,6 +4478,14 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
+  val CLIENT_BATCH_REMOVE_EXPIRED_SHUFFLE_ENABLED: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.shuffle.batchHandleRemoveExpiredShuffles.enabled")
+      .categories("client")
+      .version("0.6.0")
+      .doc("This is an optimization on remove Expired Shuffles and it's false by default.")
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4485]:
```suggestion
      .doc("Whether to batch remove expired shuffles. This is an optimization switch on removing expired shuffles.")
```

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java [hunk 29]
```diff
@@ -118,6 +118,13 @@ public ResourceResponse handleWriteRequest(ResourceProtos.ResourceRequest reques
           metaSystem.updateUnregisterShuffleMeta(shuffleKey);
           break;
 
+        case BatchUnRegisterShuffle:
+          List<String> shuffleKeys =
+              request.getBatchUnregisterShuffleRequest().getShuffleKeysList();
+          metaSystem.updateBatchUnregisterShuffleMeta(shuffleKeys);
+          LOG.debug("Handle unregister shuffle for {}", shuffleKeys);
```
Comment (SteNicholas) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java:125]:
```suggestion
          LOG.debug("Handle batch unregister shuffle for {}", shuffleKeys);
```

Comment (zaynt4606) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/MetaHandler.java:125]: Done!
Thanks~

File: common/src/main/proto/TransportMessages.proto [hunk 30]
```diff
@@ -107,6 +107,8 @@ enum MessageType {
   REPORT_BARRIER_STAGE_ATTEMPT_FAILURE_RESPONSE = 84;
   SEGMENT_START = 85;
   NOTIFY_REQUIRED_SEGMENT = 86;
+  BATCH_UNREGISTER_SHUFFLES = 87;
+  BATCH_UNREGISTER_SHUFFLE_RESPONSES= 88;
```
Comment (RexXiong) [inline common/src/main/proto/TransportMessages.proto:111]: BATCH_UNREGISTER_SHUFFLE_RESPONSES -> BATCH_UNREGISTER_SHUFFLE_RESPONSE

File: common/src/main/proto/TransportMessages.proto [hunk 31]
```diff
@@ -406,10 +408,21 @@ message PbUnregisterShuffle {
...
+  string appId = 1;
+  string requestId = 2;
+  repeated int32 shuffleIds = 3;
+}
+
 message PbUnregisterShuffleResponse {
   int32 status = 1;
 }
 
+message PbBatchUnregisterShuffleResponses {
```
Comment (RexXiong) [inline common/src/main/proto/TransportMessages.proto:421]: PbBatchUnregisterShuffleResponses -> PbBatchUnregisterShuffleResponse

File: common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala [hunk 32]
```diff
@@ -351,13 +351,36 @@ object ControlMessages extends Logging {
...
+  }
+
   object UnregisterShuffleResponse {
     def apply(status: StatusCode): PbUnregisterShuffleResponse =
       PbUnregisterShuffleResponse.newBuilder()
         .setStatus(status.getValue)
         .build()
   }
 
+  object BatchUnregisterShuffleResponses {
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala:373]: BatchUnregisterShuffleResponses -> BatchUnregisterShuffleResponse

**Review by RexXiong — APPROVED** [review]

**Comment by RexXiong** [comment]
Thanks, merge to main(v0.6.0)
